File: mobile/lib/models/change_gate.dart
enum ChangeGateStatus { idle, evaluating, evaluated, failed }

class ChangeGatePolicy {
  const ChangeGatePolicy({
    this.requirePrOpen = true,
    this.requireCiPass = true,
    this.requireHumanApproval = true,
    this.requiredApprovals = 1,
    this.allowMergeFromOneOps = false,
    this.demoApproval = false,
  });

  final bool requirePrOpen;
  final bool requireCiPass;
  final bool requireHumanApproval;
  final int requiredApprovals;
  final bool allowMergeFromOneOps;
  final bool demoApproval;

  Map<String, dynamic> toJson() {
    return {
      'requirePrOpen': requirePrOpen,
      'requireCiPass': requireCiPass,
      'requireHumanApproval': requireHumanApproval,
      'requiredApprovals': requiredApprovals,
      'allowMergeFromOneOps': allowMergeFromOneOps,
      'demoApproval': demoApproval,
    };
  }

  static const standard = ChangeGatePolicy();
  static const demoApproved = ChangeGatePolicy(demoApproval: true);
}

class ChangeGateResult {
  const ChangeGateResult({
    required this.finalState,
    required this.eligible,
    required this.reasons,
    required this.checks,
    required this.pr,
  });

  final String finalState;
  final bool eligible;
  final List<String> reasons;
  final ChangeGateChecks checks;
  final ChangeGatePullRequest pr;

  factory ChangeGateResult.fromJson(Map<String, dynamic> json) {
    return ChangeGateResult(
      finalState: _text(json['finalState'], 'BLOCKED'),
      eligible: json['eligible'] == true,
      reasons: _stringList(json['reasons']),
      checks: ChangeGateChecks.fromJson(
          (json['checks'] as Map?)?.cast<String, dynamic>() ?? const {}),
      pr: ChangeGatePullRequest.fromJson(
          (json['pr'] as Map?)?.cast<String, dynamic>() ?? const {}),
    );
  }
}

class ChangeGateChecks {
  const ChangeGateChecks({
    required this.prOpen,
    required this.targetsMain,
    required this.ciPassed,
    required this.humanApproval,
    required this.changesRequestedClear,
    required this.policySatisfied,
  });

  final bool prOpen;
  final bool targetsMain;
  final bool ciPassed;
  final bool humanApproval;
  final bool changesRequestedClear;
  final bool policySatisfied;

  factory ChangeGateChecks.fromJson(Map<String, dynamic> json) {
    return ChangeGateChecks(
      prOpen: json['prOpen'] == true,
      targetsMain: json['targetsMain'] == true,
      ciPassed: json['ciPassed'] == true,
      humanApproval: json['humanApproval'] == true,
      changesRequestedClear: json['changesRequestedClear'] == true,
      policySatisfied: json['policySatisfied'] == true,
    );
  }
}

class ChangeGatePullRequest {
  const ChangeGatePullRequest({
    required this.prNumber,
    required this.title,
    required this.state,
    required this.merged,
    required this.baseBranch,
    required this.headBranch,
    required this.url,
    required this.latestCommitSha,
    required this.reviewStatus,
    required this.reviewDecision,
    required this.approvingReviews,
    required this.requiredApprovals,
    required this.reviewPending,
    required this.changesRequested,
    required this.ciStatus,
    required this.ciSource,
    required this.changedFiles,
    required this.reviewSource,
  });

  final int prNumber;
  final String title;
  final String state;
  final bool merged;
  final String baseBranch;
  final String headBranch;
  final String url;
  final String latestCommitSha;
  final String reviewStatus;
  final String reviewDecision;
  final int approvingReviews;
  final int requiredApprovals;
  final bool reviewPending;
  final bool changesRequested;
  final String ciStatus;
  final String ciSource;
  final List<String> changedFiles;
  final String reviewSource;

  factory ChangeGatePullRequest.fromJson(Map<String, dynamic> json) {
    return ChangeGatePullRequest(
      prNumber: _int(json['prNumber']),
      title: _text(json['title'], '-'),
      state: _text(json['state'], 'unknown'),
      merged: json['merged'] == true,
      baseBranch: _text(json['baseBranch'], '-'),
      headBranch: _text(json['headBranch'], '-'),
      url: _text(json['url'], '-'),
      latestCommitSha: _text(json['latestCommitSha'], '-'),
      reviewStatus: _text(json['reviewStatus'], 'pending'),
      reviewDecision: _text(json['reviewDecision'], 'pending'),
      approvingReviews: _int(json['approvingReviews']),
      requiredApprovals: _int(json['requiredApprovals']),
      reviewPending: json['reviewPending'] == true,
      changesRequested: json['changesRequested'] == true,
      ciStatus: _text(json['ciStatus'], 'unknown'),
      ciSource: _text(json['ciSource'], 'unknown'),
      changedFiles: _stringList(json['changedFiles']),
      reviewSource: _text(json['reviewSource'], 'GitHub PR reviews'),
    );
  }
}

class ChangeGateState {
  const ChangeGateState({
    required this.status,
    this.result,
    this.error,
  });

  final ChangeGateStatus status;
  final ChangeGateResult? result;
  final String? error;

  ChangeGateState copyWith({
    ChangeGateStatus? status,
    ChangeGateResult? result,
    String? error,
    bool clearResult = false,
    bool clearError = false,
  }) {
    return ChangeGateState(
      status: status ?? this.status,
      result: clearResult ? null : result ?? this.result,
      error: clearError ? null : error ?? this.error,
    );
  }

  static const initial = ChangeGateState(status: ChangeGateStatus.idle);
}

List<String> _stringList(Object? value) {
  return ((value as List?) ?? const [])
      .map((item) => item.toString())
      .where((item) => item.trim().isNotEmpty)
      .toList();
}

String _text(Object? value, String fallback) {
  if (value == null) return fallback;
  final text = value.toString().trim();
  return text.isEmpty ? fallback : text;
}

int _int(Object? value) {
  if (value is int) return value;
  if (value is num) return value.round();
  return int.tryParse(value?.toString() ?? '') ?? 0;
}

File: backend/src/changeGate.js
const defaultPolicy = {
  requirePrOpen: true,
  requireCiPass: true,
  requireHumanApproval: true,
  requiredApprovals: 1,
  allowMergeFromOneOps: false,
  demoApproval: false,
};

export function defaultChangeGatePolicy() {
  return { ...defaultPolicy };
}

export function evaluateChangeGate({ pr, incident, policy } = {}) {
  const effectivePolicy = normalizePolicy(policy);
  const demoApproval = effectivePolicy.demoApproval === true;
  const realApprovalCount = Number(pr?.approvingReviews || 0);
  const approvalSatisfied = realApprovalCount >= effectivePolicy.requiredApprovals;
  const humanApproval = effectivePolicy.requireHumanApproval
    ? approvalSatisfied || demoApproval
    : true;
  const prOpen = effectivePolicy.requirePrOpen ? pr?.state === 'open' && pr?.merged !== true : true;
  const targetsMain = pr?.baseBranch === 'main';
  const ciPassed = effectivePolicy.requireCiPass ? pr?.ciStatus === 'passed' : true;
  const noChangesRequested = pr?.changesRequested !== true;
  const policySatisfied = prOpen && targetsMain && ciPassed && humanApproval && noChangesRequested;
  const finalState = policySatisfied
    ? 'READY_FOR_EXECUTION'
    : prOpen && targetsMain && ciPassed && noChangesRequested
      ? 'READY_FOR_APPROVAL'
      : 'BLOCKED';

  return {
    status: 'evaluated',
    finalState,
    eligible: finalState === 'READY_FOR_EXECUTION',
    reasons: buildReasons({ pr, effectivePolicy, realApprovalCount, demoApproval, prOpen, targetsMain, ciPassed, humanApproval, noChangesRequested }),
    checks: {
      prOpen,
      targetsMain,
      ciPassed,
      humanApproval,
      changesRequestedClear: noChangesRequested,
      policySatisfied,
    },
    policy: effectivePolicy,
    pr: {
      prNumber: pr?.prNumber || 0,
      title: pr?.title || '',
      state: pr?.state || 'unknown',
      merged: pr?.merged === true,
      baseBranch: pr?.baseBranch || '',
      headBranch: pr?.headBranch || '',
      url: pr?.url || '',
      latestCommitSha: pr?.latestCommitSha || '',
      reviewStatus: demoApproval && !approvalSatisfied ? 'demo_approved' : pr?.reviewStatus || 'pending',
      reviewDecision: pr?.reviewDecision || pr?.reviewStatus || 'pending',
      approvingReviews: realApprovalCount,
      requiredApprovals: effectivePolicy.requiredApprovals,
      reviewPending: pr?.reviewPending === true,
      changesRequested: pr?.changesRequested === true,
      ciStatus: pr?.ciStatus || 'unknown',
      ciSource: pr?.ciSource || 'unknown',
      ciDetails: Array.isArray(pr?.ciDetails) ? pr.ciDetails : [],
      changedFiles: Array.isArray(pr?.changedFiles) ? pr.changedFiles : [],
      reviewSource: demoApproval && !approvalSatisfied ? 'OneOps Demo Approval' : 'GitHub PR reviews',
    },
    incidentId: incident?.id || '',
  };
}

function normalizePolicy(policy = {}) {
  return {
    requirePrOpen: policy.requirePrOpen ?? defaultPolicy.requirePrOpen,
    requireCiPass: policy.requireCiPass ?? defaultPolicy.requireCiPass,
    requireHumanApproval: policy.requireHumanApproval ?? defaultPolicy.requireHumanApproval,
    requiredApprovals: Math.max(0, Number(policy.requiredApprovals ?? defaultPolicy.requiredApprovals)),
    allowMergeFromOneOps: policy.allowMergeFromOneOps === true,
    demoApproval: policy.demoApproval === true,
  };
}

function buildReasons({ pr, effectivePolicy, realApprovalCount, demoApproval, prOpen, targetsMain, ciPassed, humanApproval, noChangesRequested }) {
  const reasons = [];
  reasons.push(prOpen ? 'PR is open and unmerged' : 'PR must be open and unmerged');
  reasons.push(targetsMain ? 'PR targets main' : 'PR must target main');
  reasons.push(ciPassed ? 'CI checks passed' : `CI status is ${pr?.ciStatus || 'unknown'}`);
  if (effectivePolicy.requireHumanApproval) {
    if (humanApproval && demoApproval && realApprovalCount < effectivePolicy.requiredApprovals) {
      reasons.push('OneOps Demo Approval satisfies prototype policy only; this is not a GitHub approval');
    } else {
      reasons.push(humanApproval
        ? `${realApprovalCount} GitHub approval(s) satisfy policy`
        : `${effectivePolicy.requiredApprovals} human approval required`);
    }
  } else {
    reasons.push('Human approval is not required by policy');
  }
  reasons.push(noChangesRequested ? 'No GitHub review changes requested' : 'GitHub review changes are requested');
  if (effectivePolicy.allowMergeFromOneOps === false) {
    reasons.push('OneOps merge is disabled by policy');
  }
  return reasons;
}
